File: frontend/src/lib/api/me.ts
/**
 * Fetcher per /me sul backend FastAPI: il back end è l'unico a leggere o
 * scrivere dati applicativi (docs/adr/0001), il frontend non interroga
 * mai `utente`/`utente_privato` direttamente su Supabase. `getMe` è
 * pensata per un Server Component (usa il token già validato dal
 * layout); `completeAccount` gira invece lato client, subito dopo che
 * l'atterraggio dal link di invito ha stabilito la sessione nel browser
 * (docs/adr/0013) — non esiste ancora un giro server-side a quel punto.
 */

import type { ErroreApi } from "@/lib/api/errore";
import { ERRORE_CONFIGURAZIONE, ERRORE_RETE, erroreDaRisposta, regola } from "@/lib/api/errore";
export type IndiciStato = "pronti" | "spenti" | "in_ricostruzione";

export type Me = {
  id: string;
  nomeUtente: string;
  consensoElaborazioneAssistita: boolean;
  consensoAggiornatoAt: string;
  informativaAccettataAt: string | null;
  /** Stato osservabile della ricerca semantica (issue #6): 'pronti'
   * completa, 'spenti' nessun indice esiste, 'in_ricostruzione' il
   * lavoro in secondo piano sta ricostruendo dopo una riattivazione. */
  indiciStato: IndiciStato;
};

export type MeResult =
  | { status: "ok"; data: Me }
  // L'utente è autenticato, ma il Manutentore non ha ancora creato la
  // riga in public.utente/public.utente_privato (ADR 0007).
  | { status: "not_provisioned" }
  | { status: "error"; errore: ErroreApi };

type MeResponseBody = {
  id: string;
  nome_utente: string;
  consenso_elaborazione_assistita: boolean;
  consenso_aggiornato_at: string;
  informativa_accettata_at: string | null;
  indici_stato: IndiciStato;
};

export async function getMe(accessToken: string): Promise<MeResult> {
  const baseUrl = process.env.NEXT_PUBLIC_API_BASE_URL;
  if (!baseUrl) {
    return { status: "error", errore: ERRORE_CONFIGURAZIONE };
  }

  let response: Response;
  try {
    response = await fetch(`${baseUrl}/me`, {
      headers: { Authorization: `Bearer ${accessToken}` },
      // Dato per-sessione, mai cacheato tra richieste o utenti diversi.
      cache: "no-store",
    });
  } catch {
    return { status: "error", errore: ERRORE_RETE };
  }

  if (response.status === 404) {
    return { status: "not_provisioned" };
  }

  if (!response.ok) {
    return { status: "error", errore: erroreDaRisposta(response) };
  }

  const body = (await response.json()) as MeResponseBody;
  return { status: "ok", data: toMe(body) };
}

export type CompleteAccountResult =
  | { status: "ok"; data: Me }
  | { status: "nome_utente_in_uso" }
  // Doppio invio dello stesso completamento: non un errore per chi
  // guarda, va trattato come un successo silenzioso (l'account esiste
  // già esattamente come lo si voleva creare).
  | { status: "already_completed" }
  | { status: "validation_error"; errore: ErroreApi }
  | { status: "error"; errore: ErroreApi };

type ErrorBody = { detail?: string | { error_code?: string; message?: string } };

export async function completeAccount(
  accessToken: string,
  nomeUtente: string,
): Promise<CompleteAccountResult> {
  const baseUrl = process.env.NEXT_PUBLIC_API_BASE_URL;
  if (!baseUrl) {
    return { status: "error", errore: ERRORE_CONFIGURAZIONE };
  }

  let response: Response;
  try {
    response = await fetch(`${baseUrl}/me`, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${accessToken}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ nome_utente: nomeUtente }),
      cache: "no-store",
    });
  } catch {
    return { status: "error", errore: ERRORE_RETE };
  }

  if (response.status === 201) {
    return { status: "ok", data: toMe((await response.json()) as MeResponseBody) };
  }

  if (response.status === 422) {
    return { status: "validation_error", errore: regola("nome_utente_non_valido") };
  }

  if (response.status === 409) {
    const body = (await response.json()) as ErrorBody;
    const errorCode = typeof body.detail === "object" ? body.detail.error_code : undefined;
    if (errorCode === "nome_utente_in_uso") {
      return { status: "nome_utente_in_uso" };
    }
    if (errorCode === "account_gia_completato") {
      return { status: "already_completed" };
    }
  }

  return { status: "error", errore: erroreDaRisposta(response) };
}

function toMe(body: MeResponseBody): Me {
  return {
    id: body.id,
    nomeUtente: body.nome_utente,
    consensoElaborazioneAssistita: body.consenso_elaborazione_assistita,
    consensoAggiornatoAt: body.consenso_aggiornato_at,
    informativaAccettataAt: body.informativa_accettata_at,
    indiciStato: body.indici_stato,
  };
}

export type AggiornaConsensoResult =
  | { status: "ok"; data: Me }
  | { status: "not_provisioned" }
  | { status: "error"; errore: ErroreApi };

/**
 * Accende o spegne l'elaborazione assistita (`PATCH /me/consenso`,
 * issue #6). Un booleano solo, non un consenso per funzione: ADR 0008 ha
 * scartato la granularità perché nella pratica la decisione è una sola,
 * se i propri testi escono o no.
 *
 * Spegnendolo il backend cancella gli indici semantici; riaccendendolo li
 * fa ricostruire in secondo piano. La chiamata torna appena il flag è
 * scritto: la ricostruzione non si aspetta, la ricerca semantica dichiara
 * da sé di essere incompleta finché non è finita.
 */
export async function aggiornaConsenso(
  accessToken: string,
  consenso: boolean,
): Promise<AggiornaConsensoResult> {
  const baseUrl = process.env.NEXT_PUBLIC_API_BASE_URL;
  if (!baseUrl) {
    return { status: "error", errore: ERRORE_CONFIGURAZIONE };
  }

  let response: Response;
  try {
    response = await fetch(`${baseUrl}/me/consenso`, {
      method: "PATCH",
      headers: {
        Authorization: `Bearer ${accessToken}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ consenso }),
      cache: "no-store",
    });
  } catch {
    return { status: "error", errore: ERRORE_RETE };
  }

  if (response.status === 404) {
    return { status: "not_provisioned" };
  }

  if (!response.ok) {
    return { status: "error", errore: erroreDaRisposta(response) };
  }

  return { status: "ok", data: toMe((await response.json()) as MeResponseBody) };
}

export type EliminaAccountResult =
  | { status: "ok" }
  | { status: "conferma_non_corrispondente" }
  | { status: "not_provisioned" }
  | { status: "error"; errore: ErroreApi };

/**
 * Cancellazione self-service dell'account (`DELETE /me`, issue #8, PRD
 * regole 26-29). La conferma (il nome utente digitato) è verificata
 * server-side: questa funzione si limita a trasmetterla, non decide se è
 * corretta.
 */
export async function eliminaAccount(
  accessToken: string,
  confermaNomeUtente: string,
): Promise<EliminaAccountResult> {
  const baseUrl = process.env.NEXT_PUBLIC_API_BASE_URL;
  if (!baseUrl) {
    return { status: "error", errore: ERRORE_CONFIGURAZIONE };
  }

  let response: Response;
  try {
    response = await fetch(`${baseUrl}/me`, {
      method: "DELETE",
      headers: {
        Authorization: `Bearer ${accessToken}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ conferma_nome_utente: confermaNomeUtente }),
      cache: "no-store",
    });
  } catch {
    return { status: "error", errore: ERRORE_RETE };
  }

  if (response.status === 204) {
    return { status: "ok" };
  }

  if (response.status === 404) {
    return { status: "not_provisioned" };
  }

  if (response.status === 400) {
    const body = (await response.json()) as ErrorBody;
    const errorCode = typeof body.detail === "object" ? body.detail.error_code : undefined;
    if (errorCode === "conferma_non_corrispondente") {
      return { status: "conferma_non_corrispondente" };
    }
  }

  return { status: "error", errore: erroreDaRisposta(response) };
}

export type EsportaLibriLettiResult =
  | { status: "ok"; blob: Blob; filename: string }
  | { status: "error"; errore: ErroreApi };

const _NOME_FILE_RIPIEGO = "libri-letti.csv";

function _nomeFileDaHeader(header: string | null): string {
  if (!header) {
    return _NOME_FILE_RIPIEGO;
  }
  const match = /filename="?([^";]+)"?/.exec(header);
  return match ? match[1] : _NOME_FILE_RIPIEGO;
}

/**
 * Esportazione dei libri letti (`GET /me/export/libri-letti`, issue #8,
 * ADR 0011 rivisto): un CSV, nessuna conferma richiesta — non è
 * un'azione distruttiva. Il nome del file arriva da
 * `Content-Disposition` (esposto lato backend via CORS, app/main.py);
 * senza, si ripiega su un nome generico invece di far fallire il
 * download per un dettaglio cosmetico.
 */
export async function esportaLibriLetti(accessToken: string): Promise<EsportaLibriLettiResult> {
  const baseUrl = process.env.NEXT_PUBLIC_API_BASE_URL;
  if (!baseUrl) {
    return { status: "error", errore: ERRORE_CONFIGURAZIONE };
  }

  let response: Response;
  try {
    response = await fetch(`${baseUrl}/me/export/libri-letti`, {
      headers: { Authorization: `Bearer ${accessToken}` },
      cache: "no-store",
    });
  } catch {
    return { status: "error", errore: ERRORE_RETE };
  }

  if (!response.ok) {
    return { status: "error", errore: erroreDaRisposta(response) };
  }

  const blob = await response.blob();
  return { status: "ok", blob, filename: _nomeFileDaHeader(response.headers.get("content-disposition")) };
}

File: backend/app/routers/me.py
from fastapi import APIRouter, Depends, HTTPException, Response, status

from app.core.lingua import lingua_interfaccia
from app.core.security import get_current_user
from app.core.tempo import oggi_europa_centrale
from app.schemas.auth import AuthenticatedUser
from app.schemas.me import (
    CompleteAccountRequest,
    ConsensoUpdateRequest,
    EliminaAccountRequest,
    MeResponse,
)
from app.services import export_service, me_service

router = APIRouter(tags=["me"])


@router.get("/me", response_model=MeResponse)
async def get_me(
    current_user: AuthenticatedUser = Depends(get_current_user),  # noqa: B008
) -> MeResponse:
    result = await me_service.get_me(current_user.access_token, current_user.id)
    if result is None:
        # 404 e non 403: l'autenticazione è già validata a monte
        # (get_current_user), manca solo la risorsa applicativa, in
        # attesa che l'Utente completi l'invito (docs/adr/0013).
        raise HTTPException(
            status.HTTP_404_NOT_FOUND,
            "Il tuo account non è ancora stato completato.",
        )
    return MeResponse(**result)


@router.post("/me", response_model=MeResponse, status_code=status.HTTP_201_CREATED)
async def complete_account(
    body: CompleteAccountRequest,
    current_user: AuthenticatedUser = Depends(get_current_user),  # noqa: B008
) -> MeResponse:
    """Completa l'account dopo l'invito: sceglie `nome_utente` e registra
    l'accettazione dell'informativa nello stesso passaggio (PRD, docs/adr/0013).
    Nessun campo id/utente_id nel corpo: lo assegna il server dal token
    (AGENTS.md)."""
    try:
        result = await me_service.complete_account(current_user.access_token, body.nome_utente)
    except me_service.NomeUtenteInUsoError as error:
        raise HTTPException(
            status.HTTP_409_CONFLICT,
            {"error_code": "nome_utente_in_uso", "message": "Questo nome utente è già in uso."},
        ) from error
    except me_service.AccountGiaCompletatoError as error:
        raise HTTPException(
            status.HTTP_409_CONFLICT,
            {
                "error_code": "account_gia_completato",
                "message": "Il tuo account è già stato completato.",
            },
        ) from error
    return MeResponse(**result)


@router.patch("/me/consenso", response_model=MeResponse)
async def patch_consenso(
    body: ConsensoUpdateRequest,
    current_user: AuthenticatedUser = Depends(get_current_user),  # noqa: B008
) -> MeResponse:
    """Accende o spegne l'elaborazione assistita.

    Non `PUT /me`: `nome_utente` non è modificabile e `informativa_accettata_at`
    non deve esserlo, quindi una rotta che accettasse il profilo intero
    inviterebbe a scrivere campi che il database ora rifiuta comunque
    (grant per colonna, migrazione 20260822090000).
    """
    result = await me_service.aggiorna_consenso(
        current_user.access_token, current_user.id, body.consenso
    )
    if result is None:
        raise HTTPException(
            status.HTTP_404_NOT_FOUND,
            "Il tuo account non è ancora stato completato.",
        )
    return MeResponse(**result)


@router.get("/me/export/libri-letti")
async def esporta_libri_letti(
    current_user: AuthenticatedUser = Depends(get_current_user),  # noqa: B008
    lingua: str = Depends(lingua_interfaccia),  # noqa: B008
) -> Response:
    """CSV dei libri con stato "letto" dell'utente (issue #8, ADR 0011
    rivisto, PRD comportamento 14bis). Nessuna conferma: non è un'azione
    distruttiva."""
    contenuto = await export_service.libri_letti_csv(
        current_user.access_token, current_user.id, lingua
    )
    nome_file = f"libri-letti-{oggi_europa_centrale().isoformat()}.csv"
    return Response(
        content=contenuto,
        media_type="text/csv",
        headers={"Content-Disposition": f'attachment; filename="{nome_file}"'},
    )


@router.delete("/me", status_code=status.HTTP_204_NO_CONTENT)
async def elimina_account(
    body: EliminaAccountRequest,
    current_user: AuthenticatedUser = Depends(get_current_user),  # noqa: B008
) -> None:
    """Cancellazione self-service dell'account (issue #8, PRD regole
    26-29): la conferma è digitare il proprio nome utente, verificata
    server-side in `me_service.elimina_account` (mai fidarsi solo del
    pulsante disabilitato lato client, AGENTS.md)."""
    try:
        await me_service.elimina_account(
            current_user.access_token, current_user.id, body.conferma_nome_utente
        )
    except me_service.ContoNonTrovatoError as error:
        raise HTTPException(
            status.HTTP_404_NOT_FOUND,
            "Il tuo account non è ancora stato completato.",
        ) from error
    except me_service.ConfermaNonCorrispondenteError as error:
        raise HTTPException(
            status.HTTP_400_BAD_REQUEST,
            {
                "error_code": "conferma_non_corrispondente",
                "message": "Il nome utente digitato non corrisponde.",
            },
        ) from error
